### ISSUE TITLE
Check if the Go bindings satisfy the proposed C<->Go pointer rules

### ISSUE THREAD

**Issue by jamesr** [domokit/mojo#484]
The Go team has release proposed rules for passing pointers between C and Go code in order to remain compatible with future Go garbage collection improvements.  The proposal is here:

https://github.com/golang/proposal/blob/master/design/12416-cgo-pointers.md

referenced from go/issues/12416

We need to check if the CGo Mojo bindings satisfy these rules and, if not, what to do about it.

/cc @rogulenko @MitchRudominer @gauthamt

**Comment by jamesr** [comment]
https://github.com/domokit/mojo/blob/master/mojo/public/platform/native_cgo/system_cgo.go has most (maybe all?) of the code to review.

**Comment by MitchRudominer** [comment]
I'll take a look at this. Let me know please if anybody else is already doing it.

**Comment by MitchRudominer** [comment]
I have done the following experiment: Upgrade the version of Go in the Mojo rep to version 1.6 and set the GODEBUG environment variable to 'cgocheck=2'.  According to the [cgo doc](https://golang.org/cmd/cgo/#hdr-Passing_pointers), this enables 

> Complete checking of pointer handling, at some cost in run time

The result is that all tests are green. This either means that we have no errors in our handling of pointers or (more likely) we have insufficient test coverage. Still investigating...

**Comment by MitchRudominer** [comment]
At @qsr's suggestion I did the following experiment: I manually ran the Go Echo Client / Echo Server example Mojo app and in the Echo Server Go code I printed out the value of the GODEBUG environment variable to check that it is correctly being set in the context of multiprocess Mojo. I found that it is. When I export the variable in a bash shell and then I run
 `mojo/devtools/common/mojo_run --enable-multiprocess mojo:go_echo_client`
I find that the Go Echo Server prints out
`GODEBUG:"cgocheck=2"` as expected.

**Comment by MitchRudominer** [comment]
As far as I can tell the Mojo Go bindings do not violate the restrictions on passing pointers outlined [here](https://golang.org/cmd/cgo/#hdr-Passing_pointers).

There are several categories of restrictions and some are easier than others to see that there is no violation. One category of restriction pertains to C code calling into Go code. Since we do not have any instances of that (as far as I know) we have no violations of that category.

Focusing on Go code calling C code then there are three restrictions:
(a) Go code may not pass a pointer to Go memory if that Go memory contains a pointer to Go memory.
(b) C code may not store a Go pointer into Go memory
(c) C code may not keep a pointer to Go memory after the call to C code returns.

My methodology for analyzing whether or not any of these conditions was violated was to examine the Mojo C API and look for places where there might be a possibility of violating one of them. In doing this I cannot find any instances of a problem.

For example I considered the possibility of violating (a) when a pointer to a MojoCreateMessagePipeOptions struct that was allocated in Go memory was passed in to MojoCreateMessagePipe(). This would be a problem if a MojoCreateMessagePipeOptions struct had a field in it which was a pointer to another struct. But it does not.

I also considered the possibility of violating (c) if for example the implementation of MojoCreateMessagePipe() kept the pointer to the MojoCreateMessagePipeOptions after the call to MojoCreateMessagePipe() completed. However it is apparent from the implicit contract of the interface that it does not do that. (In other words the comments on the function would have to declare that the function took ownership of the MojoCreateMessagePipeOptions struct, and there are no such comments.)

I also looked closely at the two phase write and the two phase read in the data pipe interface as that seemed like a place where there might be an issue. But In fact there is not because those functions operate using buffers created in C memory.

Obviously this is a tricky exercise and I might be missing something. Luckily the Go runtime supplies an automated check to help give us confidence.  In the environment variable GODEBUG it is possible to set the value cgocheck to 0, 1, or 2, where 0 means no checking and 2 means maximal checking. The default value is 1.

In order to convince myself that this mechanism works I did the following exercise. I temporarily modified the interface to MojoCreateMessagePipe() so that it blatantly violated (a). I had it explicitly take a pointer to a struct that contained a pointer to another struct. I then built the Go bindings with this modification and set cgocheck=1 and I ran the Mojo Go system tests and I saw the following stack trace:
```
panic: runtime error: cgo argument has Go pointer to Go pointer

goroutine 1 [running]:
panic(0x921600, 0xc820074490)
	/home/rudominer/mojo/src/third_party/go/tool/linux_amd64/src/runtime/panic.go:464 +0x3e6
mojo/public/platform/native_cgo.(*CGoSystem).CreateMessagePipe(0xe18450, 0x0, 0x7f62b3aa24b0, 0x42b02f)
	/home/rudominer/mojo/src/mojo/public/platform/native_cgo/system_cgo.go:247 +0xfe
mojo/public/go/system.(*coreImpl).CreateMessagePipe(0xe182f0, 0x0, 0x2, 0x0, 0x0, 0x0, 0x0)
	/home/rudominer/mojo/src/mojo/public/go/system/core.go:137 +0x88
mojo/public/go/bindings.newAsyncWaiter(0xc82004fea0)
	/home/rudominer/mojo/src/mojo/public/go/bindings/async_waiter.go:227 +0x90
mojo/public/go/bindings.GetAsyncWaiter.func1()
	/home/rudominer/mojo/src/mojo/public/go/bindings/async_waiter.go:22 +0x18
sync.(*Once).Do(0xe184a0, 0xa139d8)
	/home/rudominer/mojo/src/third_party/go/tool/linux_amd64/src/sync/once.go:44 +0xe4
mojo/public/go/bindings.GetAsyncWaiter(0x0, 0x0)
	/home/rudominer/mojo/src/mojo/public/go/bindings/async_waiter.go:23 +0x3b
command-line-arguments.init.1()
	/home/rudominer/mojo/src/mojo/go/tests/testutil.go:24 +0x7b
command-line-arguments.init()
	/home/rudominer/mojo/src/mojo/go/tests/validation_type_test.go:558 +0xd0
main.init()
	command-line-arguments/_test/_testmain.go:136 +0x4a
```
I saw the same stack trace when I set cgocheck=2. I saw nothing when I set cgocheck=0.

Without my hack introducing the blatant violation I do not see any stack trace when running the system tests. This gives me some confidence that my analysis did not miss anything.
